### PR TITLE
LPS-116752 Social bookmarks share button styles

### PR DESCRIPTION
**Pull request by jbalsas**
From https://github.com/liferay-frontend/liferay-portal/pull/118, because I don't have write access in @boton's repo:

So, what we do here:
- Pass `title` attribute to `clay:dropdown-menu`
- Make sure `dynamicAttributes` are passed as props
- Exclude `locale` from `otherProps` so it won't pollute the final markup
- Remove `btn-monospace` by default server-side from all buttons that have icons (this caused fouc because it makes the label break and it's not actually present client-side)

--- 

We have two problems with share button

**Steps to reproduce:**

- Add a new widget page with the KB display widget
- Go to KB to add a basic article
- Go the page 
- View the share button 

**Expected Result:**

The share button can be displayed normally as below

image

**Actual Result:**

The share button can not be displayed normally as below

image

---

**Steps to reproduce:**

- Add a blog entry
- Add a blog widget in a page

**Expected results:**

The social bookmarks dropdown doesn't have a text and has a title.

image

**Actual results:**

The social bookmarks dropdown has a cut text and doesn't have a title.

image

Additional context from @boton:

> I pass the title to the clay dropdown here https://github.com/liferay-frontend/liferay-portal/pull/118/files#diff-f53cecdf213bf0d35836806f14a74211R83, but not render in the final HTML.
> 
> image
>
> Maybe we need to filter locale to provide clean HTML.

### CODE REVIEW

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java [hunk 1]
```diff
@@ -290,7 +290,7 @@ protected String processCssClasses(Set<String> cssClasses) {
 			cssClasses.add("btn-block");
 		}
 
-		if (Validator.isNotNull(_icon) || _monospaced) {
+		if (_monospaced) {
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:293]: Hey @pat270, what do you think about this?

I'm not sure why I added this in the first place... should all Buttons that have an icon `monospaced`? Or all buttons that **only** have an icon? 🤔 

I'm not sure about the specific impact this change might have.. 🤔

**Comment by jbalsas** [comment]
ci:test:sf

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java (hunk 1 above)
Comment (pat270) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:293]: Regular and monospaced buttons can contain icons. I think this is a good change, but looks like there will be some impact.

- modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_view_usages/page.jsp
- modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/select_asset_display_page/page.jsp
- modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/file_entry_upper_tbar.jsp
- modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/info_panel_file_entry.jsp
- modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/info_panel_file_entry.jsp
- modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/image_selector/page.jsp
- modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/page.jsp
- modules/apps/journal/journal-web/src/main/resources/META-INF/resources/article/template.jsp
- modules/apps/journal/journal-web/src/main/resources/META-INF/resources/edit_article.jsp
- modules/apps/journal/journal-web/src/main/resources/META-INF/resources/edit_ddm_template.jsp
- modules/apps/layout/layout-taglib/src/main/resources/META-INF/resources/layout_classed_model_usages_view/page.jsp
- modules/apps/portal-search/portal-search-web/src/main/resources/META-INF/resources/search/bar/view.jsp
- modules/apps/product-navigation/product-navigation-global-menu-web/src/main/resources/META-INF/resources/global_menu/global_menu.jsp
- modules/apps/ratings/ratings-taglib/src/main/resources/META-INF/resources/page.jsp


What if we did `Validator.isNotNull(_icon) && Validator.isNotNull(_label) || _monospaced`?

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:293]: > What if we did Validator.isNotNull(_icon) && Validator.isNotNull(_label) || _monospaced?

Do you mean Validator.isNotNull(_icon) && Validator.**isNull**(_label) ?

Comment (pat270) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:293]: Yes `Validator.isNotNull(_icon) && Validator.isNull(_label)`

**Comment by jbalsas** [comment]
Hey @john-co, can you take a look at these failures? 🤔

**Comment by john-co** [comment]
ci:test:bundle

**Comment by john-co** [comment]
Looks like a test fix on path locator since the html structure changed.

PR:
image

upstream:
Selection_059

I also noticed a visual regression with some icons having `">`
screenshot_18

**Comment by john-co** [comment]
I couldn't push to your liferay repo so I sent a PR in https://github.com/jbalsas/liferay-portal/pull/2222 with a possible test fix.

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by boton** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by boton** [comment]
@jbalsas It works for my two cases ✅

**Comment by jbalsas** [comment]
Thanks for verifying @boton!

@john-co, I made some adjustments and force-pushed because, honestly... I don't want to deal with more path locator madness down the road, so this change should have less impact. I'm retesting to make sure tests pass and then I'll cherry-pick your fix too, since it seems useful to prevent this from happening in the future. 

Just to provide some additional context... yesterday's code was adding unnecessary paddings, so it's good that we're changing it but functional tests should NOT be the tool to detect such things.

Finally, @drakonux, @marcoscv-work, @pat270, @eduardoallegrini, we're keeping the `btn-monospaced` class in `<clay:button>` when the button has only an icon. This causes some different alignment and sizing... so, can you check what should be the correct one?

Screen Shot 2020-07-16 at 09 06 38
Screen Shot 2020-07-16 at 09 06 46

**Comment by drakonux** [comment]
Well, @jbalsas just to say a couple of things before analyzing those classes you showed:

1. A regular input has height: **40px**. and the interactive area for the submit search button should be **40px** too.

image

2. The distance between the icon-placeholder (16x16) and the edges should be: **12px** 

padding

Checking the icon placeholder in DXP currently is (14x14)

icon-placeholder

3. The color for the search icon seems very dark probably it's using _$dark_ instead of _$secondary_

And talking about those two classes. It seems that using _btn btn-unstyled_ is closer to the final result Lexicon expects to have but, I have to say that the result has also an additional space between the button and the right edge of the input. 

margin 5px

**Comment by jbalsas** [comment]
One additional comment, is that depending on where we place the `inline-item` span, the visual output changes:

#### Now

Screen Shot 2020-07-16 at 11 55 17

#### Wrapping the label (current state of this PR)

Screen Shot 2020-07-16 at 11 55 49

#### Wrapping the icon (future state of this PR)

Screen Shot 2020-07-16 at 12 36 54

For some reason, we have a custom margin there, so I'm going to remove it and ship option 3... 🤔 

Screen Shot 2020-07-16 at 12 37 45

@drakonux, @marcoscv-work, @eduardoallegrini... thoughts?

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by pat270** [comment]
@jbalsas for the search bar classes the 2nd is correct. It should be `btn btn-unstyled`.  The correct markup:
```
<button class="btn btn-unstyled" type="submit">
  <span class="inline-item">
    <svg></svg>
  </span>
</button>
```
The reason I didn't use `btn-monospaced` in these is because having the button be 40px x 40px makes the button focus shadow appear outside the input (looks weird).
focus-shadow-outside

If we want `btn-monospaced` to work in there I can add it in Clay CSS.

**Comment by pat270** [comment]
`<span class="inline-item">` in `btn` or `btn-sm` forces `font-size: 1rem` and `btn-lg` is `font-size: 1.125rem`. The correct markup for pages tree button from a Clay CSS perspective should be:
```
<button class="btn btn-unstyled list-group-heading navigation-link panel-header-link" type="button">
  <span class="inline-item inline-item-before">
    <svg class="lexicon-icon lexicon-icon-pages-tree"></svg>
  </span>
  <span class="inline-item">Page Tree</span>
</button>
```
or
```
<button class="btn btn-unstyled list-group-heading navigation-link panel-header-link" type="button">
  <span class="inline-item inline-item-before">
    <svg class="lexicon-icon lexicon-icon-pages-tree"></svg>
  </span>
  <span>Page Tree</span>
</button>
```

**Comment by jbalsas** [comment]
Thanks @drakonux, @pat270! We can only slowly get there... so I made some small adjustments but there's more to do. @marcoscv-work, @eduardoallegrini, as we approach stabilization phase, these should be the kind of issues we should be paying attention to to make sure we polish the UI as much as possible. Please, start looking into these things to see which adjustments need to be made for the UI to be consistent and as close to our guidelines as possible!

@pat270, regarding `The correct markup for pages tree button from a Clay CSS perspective should be...`. That product team has specific css saying `font-size: 0.75rem`. Adding an `inline-item`  wrapper forces that to `1rem` which seems wrong. We might need to rethink this "correct markup". What would be the correct approach for a 0.75 rem button? `btn-sm` is `0.875rem` if I'm not mistaken... we could reach out to the team to make it sm, but I feel this is a Teams' call and we should at least have an easy way to achieve it.

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by pat270** [comment]
@jbalsas it's complex. Lexicon Design has `.btn-sm { font-size: 0.875rem; }`, but icons inside should be `1rem`. `inline-item` was how I got around the font size differences. Omitting `inline-item` is the escape hatch if you need icons in `.btn-sm` to be `0.875rem`.

For page-tree, if the icon should be `1rem` and text `0.75rem`:

```
<button class="btn btn-unstyled list-group-heading navigation-link panel-header-link" type="button">
  <span class="inline-item inline-item-before">
    <svg></svg>
  </span>
  <span>Page Tree</span>
</button>
```

if icon and text should be `0.75rem`:
```
<button class="btn btn-unstyled list-group-heading navigation-link panel-header-link" type="button">
  <span class="inline-item-before">
    <svg></svg>
  </span>
  <span>Page Tree</span>
</button>
```

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
> LocalFile.CPDocumentsandmedia#EditDocument

Looks like another case of bad locator 😭

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by john-co** [comment]
@zhangyan0701, please follow up with a test fix on this.

**Comment by john-co** [comment]
CPDocumentsandmedia#EditDocument passed...
https://testray.liferay.com/home/-/testray/case_results?testrayBuildId=1242636084

LCS test quarantined in https://github.com/brianchandotcom/liferay-portal/pull/91546

rerunning tests

**Comment by brunobasto** [comment]
Smoke test failed. That usually means flaky tests.. Running it again

**Comment by brunobasto** [comment]
ci:test:relevant

**Comment by john-co** [comment]
The LCS test can be ignored, it's been recently removed from `ci:test:relevant`.

**Comment by brunobasto** [comment]
Changes look good to me, tests are passing and according to @boton this fixes the regression bug. So I'll forward and if there's more to be done, we can always send follow up PRs.

**Comment by brunobasto** [comment]
ci:forward

**Pull request closed without merging**